system/doc/reference_manual/opaques.xml

```diff
@@ -0,0 +1,141 @@
+      <em>was</em>
+      defined in `sets` module like this:
+    </p>
+    <code type="erl">-opaque set(Element) :: #set{segs :: segs(Element)}.</code>
+    <p>OTP 24  changed the definition to the following, in
+    <url href="https://github.com/erlang/otp/commit/e66941e8d7c47b973dff94c0308ea85a6be1958e">this commit</url>
+    </p>
+    <code type="erl">-opaque set(Element) :: #set{segs :: segs(Element)} | #{Element => ?VALUE}.</code>
+    <p>
+      And this change was safer and more backwards-compatible than if the type had been defined with
+      <c>-type</c>
+      instead of
+      <c>-opaque</c>
+      . Here's why: when a module defines an
+      <c>-opaque</c>
+      , the contract is that only the defining module should rely on the definition of the type: no other modules should rely on the definition.
+    </p>
+    <p>
+      This means that code that pattern-matched on
+      <c>set</c>
+      as a record/tuple technically broke the contract, and opted in to being potentially broken when the definition of
+      <c>set()</c>
+      changed. Before OTP 24, this code printed
+      <c>ok</c>
+      . In OTP 24 it may error:
+    </p>
+    <code type="erl">
+case sets:new() of
+    Set when is_tuple(Set) -&gt;
+        io:format(&quot;ok&quot;)
+end.
+    </code>
+    <p>
+      <strong>When working with an opaque defined in another module, here are some recommendations:</strong>
+    </p>
+
+    <list type="bulleted">
+      <item>
+        Don't examine the underlying type using pattern-matching, guards, or functions that reveal the type, such as
+        <c>tuple_size/1</c>
+        .
+      </item>
+      <item>
+        Instead, use functions provided by the module for working with the type. For example,
+        <c>sets</c>
+        module provides
+        <c>sets:new/0</c>
+        ,
+        <c>sets:add/2</c>
+        ,
+        <c>sets:is_element/2</c>
+        , etc.
+      </item>
+      <item>
+        <c>sets:set(a)</c>
+        is a subtype of
+        <c>sets:set(a | b)</c>
+        and not the other way around. Generally, you can rely on the property that
+        <c>the_opaque(T)</c>
+        is a subtype of
+        <c>the_opaque(U)</c>
+        when T is a subtype of U.
+      </item>
+    </list>
+    <p>
+      <strong>When defining your own opaques, here are some recommendations:</strong>
+    </p>
+    <list type="bulleted">
+      <item>
+        Since consumers are expected to not rely on the definition of the opaque type, you must provide functions for constructing and querying/deconstructing intances of your opaque type. For example, sets can be constructed with
+        <c>sets:new/0</c>, <c>sets:from_list/1</c>, <c>sets:add/2</c>, queried with <c>sets:is_element/2</c>, and deconstructed with<c>sets:to_list/1</c>.
+      </item>
+      <item>
+        Don't define an opaque with a type variable in parameter position. This breaks the normal and expected behavior that (for example)
```

Comment:
> opaque with a type variable in parameter position

what does it mean?

Reply:
```erl
-opaque bad(T) :: fun(T) -> ok.
```

I've never seen one of these in the wild.

What's terrible about such an opaque is that the implementation leaks. `sets:set(a)` is a subtype of `sets:set(a | b)` (of course) but `bad(a)` is not a subtype of `bad(a |  b)`. Instead `bad(a | b)` is a subtype of `bad(a)`.